File: copilot-connector/sample_codes/java/common-patterns/AclConfiguration.java
// ACL configuration patterns: user, group, everyone, external groups, deny
//
// Prerequisites:
//   Maven: com.microsoft.graph:microsoft-graph:6.x, com.azure:azure-identity:1.x
//   Gradle: implementation 'com.microsoft.graph:microsoft-graph:6.+'
//           implementation 'com.azure:azure-identity:1.+'

import com.microsoft.graph.GraphServiceClient;
import com.microsoft.graph.models.externalconnectors.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Demonstrates all ACL patterns for Copilot Connector items.
 * Covers public, internal-only, user-restricted, group-restricted,
 * everyone-except (deny), external group ACLs, and external group management.
 */
public class AclConfiguration {

    private final GraphServiceClient graphClient;
    private final String connectionId;

    public AclConfiguration(GraphServiceClient graphClient, String connectionId) {
        this.graphClient = graphClient;
        this.connectionId = connectionId;
    }

    /**
     * Public item visible to all tenant users.
     */
    public List<Acl> publicAcl() {
        Acl acl = new Acl();
        acl.setType(AclType.Everyone);
        acl.setValue("everyone");
        acl.setAccessType(AccessType.Grant);
        return List.of(acl);
    }

    /**
     * Internal item visible to all tenant users except guests.
     */
    public List<Acl> internalOnlyAcl() {
        Acl acl = new Acl();
        acl.setType(AclType.EveryoneExceptGuests);
        acl.setValue("everyoneExceptGuests");
        acl.setAccessType(AccessType.Grant);
        return List.of(acl);
    }

    /**
     * Item restricted to specific Entra ID users.
     * Values MUST be Entra object IDs (GUIDs), not emails or UPNs.
     *
     * @param userEntraIds one or more Entra ID user GUIDs
     */
    public List<Acl> userRestrictedAcl(String... userEntraIds) {
        return Arrays.stream(userEntraIds)
                .map(id -> {
                    Acl acl = new Acl();
                    acl.setType(AclType.User);
                    acl.setValue(id);
                    acl.setAccessType(AccessType.Grant);
                    return acl;
                })
                .collect(Collectors.toList());
    }

    /**
     * Item restricted to members of specific Entra ID groups.
     * Prefer group-based ACLs over user-based for maintainability.
     *
     * @param groupEntraIds one or more Entra ID group GUIDs
     */
    public List<Acl> groupRestrictedAcl(String... groupEntraIds) {
        return Arrays.stream(groupEntraIds)
                .map(id -> {
                    Acl acl = new Acl();
                    acl.setType(AclType.Group);
                    acl.setValue(id);
                    acl.setAccessType(AccessType.Grant);
                    return acl;
                })
                .collect(Collectors.toList());
    }

    /**
     * Combination: grant to all, deny to specific users.
     * Deny ALWAYS overrides grant.
     *
     * @param deniedUserIds Entra ID GUIDs of users to deny
     */
    public List<Acl> everyoneExceptAcl(String... deniedUserIds) {
        var aclList = new ArrayList<Acl>();

        Acl grantAll = new Acl();
        grantAll.setType(AclType.Everyone);
        grantAll.setValue("everyone");
        grantAll.setAccessType(AccessType.Grant);
        aclList.add(grantAll);

        for (String id : deniedUserIds) {
            Acl deny = new Acl();
            deny.setType(AclType.User);
            deny.setValue(id);
            deny.setAccessType(AccessType.Deny);
            aclList.add(deny);
        }

        return aclList;
    }

    /**
     * Item restricted to an external group (for non-Entra ID permissions).
     * External groups must be created first via the group sync API.
     *
     * @param externalGroupId the external group ID
     */
    public List<Acl> externalGroupAcl(String externalGroupId) {
        Acl acl = new Acl();
        acl.setType(AclType.ExternalGroup);
        acl.setValue(externalGroupId);
        acl.setAccessType(AccessType.Grant);
        return List.of(acl);
    }

    // --- External Group Management ---

    /**
     * Create an external group for non-Entra ID permission structures.
     * Use for Salesforce roles, ServiceNow groups, custom RBAC, etc.
     *
     * @param groupId     unique ID for the external group
     * @param displayName human-readable name
     * @param description optional description
     * @return the created external group
     */
    public ExternalGroup createExternalGroup(
            String groupId, String displayName, String description) {

        ExternalGroup group = new ExternalGroup();
        group.setId(groupId);
        group.setDisplayName(displayName);
        group.setDescription(description);

        ExternalGroup created = graphClient.external().connections()
                .byExternalConnectionId(connectionId)
                .groups()
                .post(group);

        if (created == null) {
            throw new RuntimeException("Failed to create external group");
        }
        return created;
    }

    /**
     * Add an Entra ID user to an external group.
     *
     * @param groupId     the external group ID
     * @param userEntraId the Entra ID (GUID) of the user
     */
    public void addUserToExternalGroup(String groupId, String userEntraId) {
        Identity member = new Identity();
        member.setOdataType("#microsoft.graph.externalConnectors.identity");
        member.setId(userEntraId);
        member.setType(IdentityType.User);

        graphClient.external().connections()
                .byExternalConnectionId(connectionId)
                .groups()
                .byExternalGroupId(groupId)
                .members()
                .post(member);
    }

    /**
     * Add an Entra ID group to an external group (nested groups).
     *
     * @param externalGroupId the external group ID
     * @param entraGroupId    the Entra ID (GUID) of the group to add
     */
    public void addGroupToExternalGroup(String externalGroupId, String entraGroupId) {
        Identity member = new Identity();
        member.setOdataType("#microsoft.graph.externalConnectors.identity");
        member.setId(entraGroupId);
        member.setType(IdentityType.Group);

        graphClient.external().connections()
                .byExternalConnectionId(connectionId)
                .groups()
                .byExternalGroupId(externalGroupId)
                .members()
                .post(member);
    }
}

File: copilot-connector/sample_codes/python/common-patterns/acl-configuration.py
# ACL configuration patterns: user, group, everyone, external groups, deny
# Prerequisites: pip install msgraph-sdk azure-identity

from msgraph import GraphServiceClient
from msgraph.generated.models.external_connectors.acl import Acl
from msgraph.generated.models.external_connectors.acl_type import AclType
from msgraph.generated.models.external_connectors.access_type import AccessType
from msgraph.generated.models.external_connectors.external_group import ExternalGroup
from msgraph.generated.models.external_connectors.identity import Identity
from msgraph.generated.models.external_connectors.identity_type import IdentityType


class AclConfiguration:
    """ACL configuration patterns for Copilot Connector items."""

    def __init__(self, graph_client: GraphServiceClient, connection_id: str) -> None:
        self._graph_client = graph_client
        self._connection_id = connection_id

    # --- ACL Patterns ---

    @staticmethod
    def public_acl() -> list[Acl]:
        """Public item visible to all tenant users."""
        return [
            Acl(
                type=AclType.Everyone,
                value="everyone",
                access_type=AccessType.Grant,
            )
        ]

    @staticmethod
    def internal_only_acl() -> list[Acl]:
        """Public item visible to all tenant users except guests."""
        return [
            Acl(
                type=AclType.EveryoneExceptGuests,
                value="everyoneExceptGuests",
                access_type=AccessType.Grant,
            )
        ]

    @staticmethod
    def user_restricted_acl(*user_entra_ids: str) -> list[Acl]:
        """
        Item restricted to specific Entra ID users.
        Values MUST be Entra object IDs (GUIDs), not emails or UPNs.
        """
        return [
            Acl(
                type=AclType.User,
                value=user_id,
                access_type=AccessType.Grant,
            )
            for user_id in user_entra_ids
        ]

    @staticmethod
    def group_restricted_acl(*group_entra_ids: str) -> list[Acl]:
        """
        Item restricted to members of specific Entra ID groups.
        Prefer group-based ACLs over user-based for maintainability.
        """
        return [
            Acl(
                type=AclType.Group,
                value=group_id,
                access_type=AccessType.Grant,
            )
            for group_id in group_entra_ids
        ]

    @staticmethod
    def everyone_except_acl(*denied_user_ids: str) -> list[Acl]:
        """
        Combination: grant to all, deny to specific users.
        Deny ALWAYS overrides grant.
        """
        acl: list[Acl] = [
            Acl(
                type=AclType.Everyone,
                value="everyone",
                access_type=AccessType.Grant,
            )
        ]

        acl.extend(
            Acl(
                type=AclType.User,
                value=user_id,
                access_type=AccessType.Deny,
            )
            for user_id in denied_user_ids
        )

        return acl

    @staticmethod
    def external_group_acl(external_group_id: str) -> list[Acl]:
        """
        Item restricted to an external group (for non-Entra ID permissions).
        External groups must be created first via the group sync API.
        """
        return [
            Acl(
                type=AclType.ExternalGroup,
                value=external_group_id,
                access_type=AccessType.Grant,
            )
        ]

    # --- External Group Management ---

    async def create_external_group(
        self,
        group_id: str,
        display_name: str,
        description: str | None = None,
    ) -> ExternalGroup:
        """
        Create an external group for non-Entra ID permission structures.
        Use for Salesforce roles, ServiceNow groups, custom RBAC, etc.
        """
        group = ExternalGroup(
            id=group_id,
            display_name=display_name,
            description=description,
        )

        result = await self._graph_client.external.connections.by_external_connection_id(
            self._connection_id
        ).groups.post(group)

        if result is None:
            raise RuntimeError("Failed to create external group")
        return result

    async def add_user_to_external_group(
        self, group_id: str, user_entra_id: str
    ) -> None:
        """Add an Entra ID user to an external group."""
        member = Identity(
            odata_type="#microsoft.graph.externalConnectors.identity",
            id=user_entra_id,
            type=IdentityType.User,
        )

        await self._graph_client.external.connections.by_external_connection_id(
            self._connection_id
        ).groups.by_external_group_id(group_id).members.post(member)

    async def add_group_to_external_group(
        self, external_group_id: str, entra_group_id: str
    ) -> None:
        """Add an Entra ID group to an external group (nested groups)."""
        member = Identity(
            odata_type="#microsoft.graph.externalConnectors.identity",
            id=entra_group_id,
            type=IdentityType.Group,
        )

        await self._graph_client.external.connections.by_external_connection_id(
            self._connection_id
        ).groups.by_external_group_id(external_group_id).members.post(member)
